src/interface/state_descriptor.cpp

```diff
@@ -238,40 +231,52 @@ bool StateDescriptor::AddSwarmValue(const std::string &value_name,
   return true;
 }
 
-bool StateDescriptor::AddField(const std::string &field_name, const Metadata &m_in) {
+bool StateDescriptor::AddFieldImpl(const VarID &vid, const Metadata &m_in) {
   Metadata m = m_in; // Force const correctness
-  if (m.IsSet(Metadata::Sparse)) {
-    int sparse_id = m.GetSparseId();
-    if (sparseMetadataMap_.count(field_name) > 0) {
-      if (sparseMetadataMap_.at(field_name).count(sparse_id) > 0) {
-        return false;
-      }
-    }
-    sparseMetadataMap_[field_name][sparse_id] = m;
+
+  const std::string &assoc = m.getAssociated();
+  if (m.getAssociated() == "") {
+    m.Associate(vid.label());
+  }
+  if (FieldPresent(vid.label()) || SparseBaseNamePresent(vid.label())) {
+    return false; // this field has already been added
   } else {
-    const std::string &assoc = m.getAssociated();
-    if (!assoc.length()) m.Associate(field_name);
-    if (metadataMap_.count(field_name) > 0) {
-      return false; // this field has already been added
-    } else {
-      metadataMap_[field_name] = m;
-    }
+    metadataMap_.insert({vid, m});
   }
+
   return true;
 }
 
-bool StateDescriptor::FlagsPresent(std::vector<MetadataFlag> const &flags,
-                                   bool matchAny) {
-  for (auto &pair : metadataMap_) {
-    auto &metadata = pair.second;
-    if (metadata.FlagsSet(flags, matchAny)) return true;
+bool StateDescriptor::AddSparsePoolImpl(const SparsePool &pool) {
+  if (pool.pool().size() == 0) {
+    return false;
   }
-  for (auto &pair : sparseMetadataMap_) {
-    auto &sparsemap = pair.second;
-    for (auto &p2 : sparsemap) {
-      if (p2.second.FlagsSet(flags, matchAny)) return true;
-    }
+
+  if (FieldPresent(pool.base_name()) || SparseBaseNamePresent(pool.base_name())) {
+    // this sparse variable has already been added
+    return false;
   }
+
+  sparsePoolMap_.insert({pool.base_name(), pool});
+
+  // add all the sparse fields
+  for (const auto itr : pool.pool()) {
+    AddFieldImpl(VarID(pool.base_name(), itr.first), itr.second);
+  }
+
+  return true;
+}
```

Comment:
Good question. I'm actually not familiar with the associated machinery and use cases, so I could use some help there to understand what the needs and intentions are.

Right now, the associated field is shared between all the fields in a sparse pool. So if you create a sparse pool called "sparse_A" with associated "sparse_B", then all the sparse IDs in "sparse_A" will have associated "sparse_B". It would be easy to change this so that the sparse ID X in "sparse_A" has associated "sparse_B_X", but there is no guarantee that the "sparse_B" pool also has a sparse ID X. Adding such a check to `StateDescriptor::CreateResolvedStateDescriptor` should be possible and fairly straight forward, though.

Reply:
Hmm. Ok. I *think* what you have now is the right behaviour. I'm trying to think of a case where you would want to associate a sparse variable with ID X to a different sparse variable with ID Y, and I can't currently think of one.

Maybe the thing to do is just document this behaviour somewhere and be aware of it. I don't think it would be a big lift to change it later if needed.

@jdolence do you have thoughts?

Reply:
I've added this to `state.md` documentation.